### PR TITLE
1234changes

### PR DESCRIPTION
**Pull request by aadiegoaa96**
a lot of changes for fix my format, practices like change then for map, use of verification jwt token like: 

jwt.verify(token, SECRET_KEY, async (err, decoded) => {

### CODE REVIEW

File: app.ts [hunk 1]
```diff
@@ -2,47 +2,37 @@ import express, { Request, Response, NextFunction } from 'express';
...
-// Llamada a la función para generar el token JWT
 generateJWT();
 
-// Conectar a la base de datos de MongoDB
 mongoose
   .connect(MONGO_URL, {
     useNewUrlParser: true,
     useUnifiedTopology: true
   } as ConnectOptions)
   .then(async () => {
```
Comment (danielavaladezacc) [inline app.ts:35]: Change this .then to use an await/async

**Review by danielavaladezacc — CHANGES_REQUESTED** [review]
Leave the handler as simple as possible and put all of the orchestration logic in the controller.

File: app.ts (hunk 1 above)
Comment (aadiegoaa96) [reply app.ts]: `import express, { Request, Response, NextFunction } from 'express';
import mongoose, { ConnectOptions } from 'mongoose';
import { routes } from './routes/routes';
import jwt from 'jsonwebtoken';

const app = express();
const port = 3000;
const SECRET_KEY = 'gauss626';
const MONGO_URL = 'mongodb://127.0.0.1:27017/mydatabase';
const COLLECTION_NAME = 'log-collection';

app.use(express.json());

app.use(async (req: Request, res: Response, next: NextFunction) => {
  // ... código de middleware para guardar el log ...
  console.log('Server started');
  next();
});

app.use('/', routes);

// Generar el token JWT y mostrarlo en la consola
const generateJWT = () => {
  const payload = {
    username: 'example',
    role: 'admin'
  };

  const token = jwt.sign(payload, SECRET_KEY);
  console.log('Token JWT:', token);
};

// Llamada a la función para generar el token JWT
generateJWT();

// Conectar a la base de datos de MongoDB
mongoose
  .connect(MONGO_URL, {
    useNewUrlParser: true,
    useUnifiedTopology: true
  } as ConnectOptions)
  .then(async () => {
    // ... código para crear la colección "log-collection" ...

    // Configurar la aplicación Express para escuchar en el puerto especificado
    app.listen(port, () => {
      console.log(`Servidor escuchando en http://localhost:${port}`);
    });
  })
  .catch((error) => {
    console.error('Error al conectar a MongoDB:', error);
  });

export { app };`

File: controller/controller.ts [hunk 2]
```diff
@@ -1,9 +1,9 @@
+import transformUser from '../utils/utils';
+import Handler from '../handler/handler';
 import axios from 'axios';
 import asyncRetry from 'async-retry';
-import { User } from './models';
-import { transformUser } from './utils';
 
-const getUsers = async (): Promise<User[]> => {
+const getUsers = async (): Promise<any[]> => {
```
[the remark this answers is not in the document]
Comment (aadiegoaa96) [reply controller/controller.ts]: `import { transformUser } from '../utils/utils';
import { getUsers } from '../services/services';
import { User } from '../models/models';

const getUsersController = async (): Promise<User[]> => {
  try {
    const users = await getUsers();
    const transformedUsers = users.map(transformUser);
    return transformedUsers;
  } catch (error) {
    console.error('Error al llamar a la API:', error);
    throw error;
  }
};

export { getUsersController };`

File: controller/controller.ts [hunk 3]
```diff
@@ -1,9 +1,9 @@
...
   try {
     const response = await asyncRetry(() => axios.get('https://jsonplaceholder.typicode.com/users'));
     const transformedUsers = response.data.map(transformUser);
```
[the remark this answers is not in the document]
Comment (aadiegoaa96) [reply controller/controller.ts]: `import { transformUser } from '../utils/utils';
import { getUsers } from '../services/services';
import { User } from '../models/models';

const getUsersController = async (): Promise<User[]> => {
  try {
    const users = await getUsers();
    const transformedUsers = users.map(transformUser);
    return transformedUsers;
  } catch (error) {
    console.error('Error al llamar a la API:', error);
    throw error;
  }
};

export { getUsersController };`

File: handler/handler.ts [hunk 4]
```diff
@@ -1,28 +1,39 @@
...
+      username: 'example',
+      role: 'admin'
+    };
+    const token = jwt.sign(payload, SECRET_KEY);
+    res.json({ token });
+  }
+}
 
-export { getUsersHandler, loginHandler };
+export default new Handler();
```
[the remark this answers is not in the document]
Comment (aadiegoaa96) [reply handler/handler.ts]: `import { Request, Response } from 'express';
import jwt from 'jsonwebtoken';
import { getUsersController } from '../controller/controller';
import { User } from '../models/models';

const SECRET_KEY = 'gauss626';

export class Handler {
  async getUsersHandler(req: Request, res: Response) {
    try {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) {
        return res.status(401).json({ message: 'No se proporcionó un token de autenticación' });
      }

      jwt.verify(token, SECRET_KEY, async (err, decoded) => {
        if (err) {
          return res.status(401).json({ message: 'Token de autenticación inválido' });
        }

        const transformedUsers: User[] = await getUsersController();
        res.json(transformedUsers);
      });
    } catch (error) {
      console.error('Error al llamar a la API:', error);
      res.sendStatus(500);
    }
  }

  loginHandler(req: Request, res: Response) {
    const payload = {
      username: 'example',
      role: 'admin'
    };
    const token = jwt.sign(payload, SECRET_KEY);
    res.json({ token });
  }
}`

File: services/services.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import axios from 'axios';
+
+class UserService {
+  async getUsers() {
```
[the remark this answers is not in the document]
Comment (aadiegoaa96) [reply services/services.ts]: `import axios from 'axios';
import { User } from '../models/models';

export const getUsers = async (): Promise<User[]> => {
  try {
    const response = await axios.get('https://jsonplaceholder.typicode.com/users');
    return response.data;
  } catch (error) {
    console.error('Error al llamar a la API:', error);
    throw error;
  }
};`

File: services/services.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+      const response = await axios.get('https://jsonplaceholder.typicode.com/users');
+      return response.data;
+    } catch (error) {
+      console.error('Error al llamar a la API:', error);
+      throw error;
+    }
+  }
+}
+
+export default new UserService();
```
[the remark this answers is not in the document]
Comment (aadiegoaa96) [reply services/services.ts]: `import axios from 'axios';
import { User } from '../models/models';

export const getUsers = async (): Promise<User[]> => {
  try {
    const response = await axios.get('https://jsonplaceholder.typicode.com/users');
    return response.data;
  } catch (error) {
    console.error('Error al llamar a la API:', error);
    throw error;
  }
};`

File: utils/utils.ts [hunk 7]
```diff
@@ -9,4 +9,4 @@ const transformUser = (user: any): User => ({
   body: '',
 });
 
-export { transformUser };
+export default transformUser;
```
[the remark this answers is not in the document]
Comment (aadiegoaa96) [reply utils/utils.ts]: `import { User } from '../models/models';

export const transformUser = (user: any): User => ({
  userId: user.id,
  name: user.name,
  email: user.email,
  postId: '',
  title: '',
  body: ''
});`

File: handler/handler.ts [hunk 8]
```diff
@@ -1,28 +1,39 @@
...
+          return res.status(401).json({ message: 'Token de autenticación inválido' });
+        }
 
-const loginHandler = (req: Request, res: Response) => {
-  // Generar y devolver el token JWT en la respuesta
-  const payload = {
-    username: 'example',
-    role: 'admin'
-  };
+        const transformedUsers = await UserService.getUsers();
```
[the remark this answers is not in the document]
Comment (aadiegoaa96) [reply handler/handler.ts]: `import { Request, Response } from 'express';
import jwt from 'jsonwebtoken';
import { getUsersController } from '../controller/controller';
import { User } from '../models/models';

const SECRET_KEY = 'gauss626';

export class Handler {
  async getUsersHandler(req: Request, res: Response) {
    try {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) {
        return res.status(401).json({ message: 'No se proporcionó un token de autenticación' });
      }

      jwt.verify(token, SECRET_KEY, async (err, decoded) => {
        if (err) {
          return res.status(401).json({ message: 'Token de autenticación inválido' });
        }

        const transformedUsers: User[] = await getUsersController();
        res.json(transformedUsers);
      });
    } catch (error) {
      console.error('Error al llamar a la API:', error);
      res.sendStatus(500);
    }
  }

  loginHandler(req: Request, res: Response) {
    const payload = {
      username: 'example',
      role: 'admin'
    };
    const token = jwt.sign(payload, SECRET_KEY);
    res.json({ token });
  }
}`

File: controller/controller.ts [hunk 9]
```diff
@@ -1,17 +1,16 @@
...
+import { getUsers } from '../services/services';
+import { User } from '../models/models';
 
-const getUsers = async (): Promise<User[]> => {
+const getUsersController = async (): Promise<User[]> => {
   try {
-    const response = await asyncRetry(() => axios.get('https://jsonplaceholder.typicode.com/users'));
-    const transformedUsers = response.data.map(transformUser);
+    const users = await getUsers();
+    const transformedUsers = users.map(transformUser);
```
Comment (danielavaladezacc) [inline controller/controller.ts:8]: This is still not transforming the users into the expected format. The `map` function here is not doing anything. Also, remember to move this to a `util` file.

File: controller/controller.ts [hunk 10]
```diff
@@ -1,17 +1,16 @@
-import axios from 'axios';
-import asyncRetry from 'async-retry';
-import { User } from './models';
-import { transformUser } from './utils';
+import { transformUser } from '../utils/utils';
+import { getUsers } from '../services/services';
+import { User } from '../models/models';
 
-const getUsers = async (): Promise<User[]> => {
+const getUsersController = async (): Promise<User[]> => {
```
Comment (danielavaladezacc) [inline controller/controller.ts:5]: A controller should be either a class with methods or a constant with various functions. This is just one single function.

File: controller/controller.ts [hunk 11]
```diff
@@ -1,17 +1,16 @@
...
+    const transformedUsers = users.map(transformUser);
     return transformedUsers;
   } catch (error) {
     console.error('Error al llamar a la API:', error);
     throw error;
   }
 };
 
-export { getUsers };
+export { getUsersController };
```
Comment (danielavaladezacc) [inline controller/controller.ts:16]: Remove this line. Export the class only.

File: services/services.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import axios from 'axios';
+import { User } from '../models/models';
+
+export const getUsers = async (): Promise<User[]> => {
```
Comment (danielavaladezacc) [inline services/services.ts:4]: Same like with the controller, this either needs to be a class with methods or a constant with functions in them.

File: utils/utils.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import { User } from '../models/models';
+
+export const transformUser = (user: any): User => ({
```
Comment (danielavaladezacc) [inline utils/utils.ts:3]: Same as service and controller. Don't just export a single function.

File: utils/utils.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+  userId: user.id,
+  name: user.name,
+  email: user.email,
+  postId: '',
```
Comment (danielavaladezacc) [inline utils/utils.ts:7]: This is to transform the users' posts, not a user. Also, I don't see this being called anywhere in the code?

**Review by danielavaladezacc — CHANGES_REQUESTED** [review]

File: utils/utils.ts (hunk 14 above)
Comment (aadiegoaa96) [reply utils/utils.ts]: getUserPosts
  
  it's called from the controller.ts method
  
  `  async getUserPosts(userId: string): Promise<TransformedPosts[]> {
    try {
      const posts = await getPostsByUserId(userId);
      const transformedPosts = usersUtil.transformUserPosts(posts);

      return transformedPosts;
    } catch (error) {
      console.error('Error calling API:', error);
      throw error;
    }
  }`

File: controller/controller.ts (hunk 9 above)
Comment (aadiegoaa96) [reply controller/controller.ts]: I change the .map to utils, and I fix the format

export const usersUtil = {
  transformUsers(users: User[]): TransformedUser[] {
    return users.map((user) => {
      const nameParts = user.name.split(' '); // Dividir el nombre en partes
      const firstName = nameParts.slice(validPrefixes.includes(nameParts[0]) ? 1 : 0, -1).join(' '); // Unir los nombres restantes en el primer nombre, excluyendo el apellido
      const prefix = validPrefixes.includes(nameParts[0]) ? nameParts[0] : ''; // Obtener el prefijo si existe, de lo contrario, dejarlo vacío
      const lastName = nameParts[nameParts.length - 1] || ''; // Obtener el apellido si existe, de lo contrario, dejarlo vacío
      const companyName = user.company.name || ''; // Obtener el nombre de la compañía si existe, de lo contrario, dejarlo vacío

      const transformedUser: TransformedUser = {
        id: user.id,
        prefix,
        firstName,
        lastName,
        email: user.email,
        address: `${user.address.street} ${user.address.suite} ${user.address.city} ${user.address.zipcode}`,
        geolocation: `(${user.address.geo.lat}, ${user.address.geo.lng})`,
        companyName
      };

      return transformedUser;
    });
  },

  transformUserPosts(posts: Post[]): TransformedPosts[] {
    return posts.map((post) => {
      return {
        postId: post.postId,
        userId: post.userId,
        title: post.title,
        body: post.body
      };
    });
  }
};

File: controller/controller.ts (hunk 11 above)
Comment (aadiegoaa96) [reply controller/controller.ts]: It's ok? 

`// >> controller\controller.ts

import { usersUtil } from '../utils/utils';
import { getUsers, getPostsByUserId } from '../services/services';
import { User, Post, TransformedUser, TransformedPosts } from '../models/models';

export class UsersController {
  constructor() {}

  async getUsers(): Promise<TransformedUser[]> {
    try {
      const users = await getUsers();
      const transformedUsers = usersUtil.transformUsers(users);

      return transformedUsers;
    } catch (error) {
      console.error('Error calling API:', error);
      throw error;
    }
  }

  async getUserPosts(userId: string): Promise<TransformedPosts[]> {
    try {
      const posts = await getPostsByUserId(userId);
      const transformedPosts = usersUtil.transformUserPosts(posts);

      return transformedPosts;
    } catch (error) {
      console.error('Error calling API:', error);
      throw error;
    }
  }
}`

File: services/services.ts (hunk 12 above)
Comment (aadiegoaa96) [reply services/services.ts]: this is my services.ts file with classes: 

`//services/services.ts
import axios from 'axios';
import { User, Post } from '../models/models';

export class UserService {
  static async getUsers(): Promise<User[]> {
    try {
      const response = await axios.get('https://jsonplaceholder.typicode.com/users');
      return response.data;
    } catch (error) {
      console.error('Error calling API:', error);
      throw error;
    }
  }
}

export class PostService {
  static async getPostsByUserId(userId: string): Promise<Post[]> {
    try {
      const response = await axios.get(`https://jsonplaceholder.typicode.com/posts?userId=${userId}`);
      return response.data;
    } catch (error) {
      console.error('Error calling API:', error);
      throw error;
    }
  }
}
`

**Pull request merged**